test/integration/health_check_integration_test.cc

```diff
@@ -0,0 +1,261 @@
+    initialize();
+    initHttpHealthCheck(cluster_idx);
+
+    // Endpoint responds with healthy status to the health check.
+    clusters_[cluster_idx].host_stream_->encodeHeaders(
+        Http::TestResponseHeaderMapImpl{{":status", "200"}}, false);
+    clusters_[cluster_idx].host_stream_->encodeData(1024, true);
+
+    // Verify that Envoy detected the health check response.
+    test_server_->waitForCounterGe("cluster.cluster_1.health_check.success", 1);
+    EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.success")->value());
+    EXPECT_EQ(0, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
+
+    // Clean up connections.
+    cleanupHostConnections();
+  }
+
+  // Tests that an unhealthy endpoint returns a valid HTTP health check response.
+  // HTTP tests are defined here to check both HTTP1 and HTTP2.
+  void singleEndpointUnhealthyHttp(envoy::type::v3::CodecClientType codec_client_type) {
+    codec_client_type_ = codec_client_type;
+
+    const uint32_t cluster_idx = 0;
+    initialize();
+    initHttpHealthCheck(cluster_idx);
+
+    // Endpoint responds to the health check with unhealthy status.
+    clusters_[cluster_idx].host_stream_->encodeHeaders(
+        Http::TestResponseHeaderMapImpl{{":status", "503"}}, false);
+    clusters_[cluster_idx].host_stream_->encodeData(1024, true);
+
+    test_server_->waitForCounterGe("cluster.cluster_1.health_check.failure", 1);
+    EXPECT_EQ(0, test_server_->counter("cluster.cluster_1.health_check.success")->value());
+    EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
+
+    // Clean up connections.
+    cleanupHostConnections();
+  }
+
+  // Tests that no HTTP health check response results in timeout and unhealthy endpoint.
+  // HTTP tests are defined here to check both HTTP1 and HTTP2.
+  void singleEndpointTimeoutHttp(envoy::type::v3::CodecClientType codec_client_type) {
+    codec_client_type_ = codec_client_type;
+
+    const uint32_t cluster_idx = 0;
+    initialize();
+    // Set timeout to 1 second, so the waiting will not be long.
+    initHttpHealthCheck(cluster_idx, 1);
+
+    // Endpoint doesn't reply, and a healthcheck failure occurs (due to timeout).
+    test_server_->waitForCounterGe("cluster.cluster_1.health_check.failure", 1);
+    EXPECT_EQ(0, test_server_->counter("cluster.cluster_1.health_check.success")->value());
+    EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
+
+    // Clean up connections
+    cleanupHostConnections();
+  }
+
+  // The number of clusters and their names must match the clusters in the CDS integration test
+  // configuration.
+  static constexpr size_t clusters_num_ = 2;
+  std::array<ClusterData, clusters_num_> clusters_{{{"cluster_1"}, {"cluster_2"}}};
+  envoy::type::v3::CodecClientType codec_client_type_{envoy::type::v3::HTTP2};
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, HealthCheckIntegrationTest,
```

Comment:
do you expect all or most tests to be parametrized on codec type? if so, maybe use a param for `codec_client_type_`?

Reply:
The others will be TCP (no codec), and gRPC (only HTTP2).
Only the HTTP tests will be "duplicated" for HTTP/1 and HTTP/2.

Reply:
Ah, gotcha! Sounds good, I though maybe this was a base class and TCP and gRPC would be in different classes.

Reply:
I've refactored the test suite to use 3 different classes (for HTTP, TCP, and gRPC).
Now there's an `upstream_protocol_` that specifies the required type when creating the host upstream.